File: ELEC881_Tabnine/src/No Description Prompt/Java script (JS)/q29-longest-increasing-path-in-a-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number}
 */
var longestIncreasingPath = function(matrix) {
    if (!matrix.length ||!matrix[0].length) return 0;
    const m = matrix.length;
    const n = matrix[0].length;
    const dp = Array.from({ length: m }, () => Array.from({ length: n }, () => 0));
    let maxLen = 0;
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            maxLen = Math.max(maxLen, dfs(i, j, matrix, dp));
        }
    }
    return maxLen;
};

function dfs(i, j, matrix, dp) {
    if (dp[i][j] > 0) return dp[i][j];
    let maxLen = 1;
    const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    for (const [di, dj] of directions) {
        const ni = i + di;
        const nj = j + dj;
        if (
            ni >= 0 && ni < m && nj >= 0 && nj < n &&
            matrix[ni][nj] < matrix[i][j]
        ) {
            maxLen = Math.max(maxLen, dfs(ni, nj, matrix, dp) + 1);
        }
    }
    dp[i][j] = maxLen;
    return maxLen;

}

File: ELEC881_Tabnine/src/No Description Prompt/Python/q29-longest-increasing-path-in-a-matrix.py
class Solution(object):
    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        if not matrix or not matrix[0]:
            return 0

        m, n = len(matrix), len(matrix[0])
        dp = [[0] * n for _ in range(m)]
        res = 0
        for i in range(m):
            for j in range(n):
                dp[i][j] = self.dfs(matrix, i, j, dp)
                res = max(res, dp[i][j])
                return res

    def dfs(self, matrix, i, j, dp):
        if dp[i][j] > 0:
            return dp[i][j]

        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        max_len = 1
        for dx, dy in directions:
            x, y = i + dx, j + dy
            if 0 <= x < len(matrix) and 0 <= y < len(matrix[0]) and matrix[x][y] > matrix[i][j]:
                max_len = max(max_len, self.dfs(matrix, x, y, dp) + 1)

        dp[i][j] = max_len
        return max_len



        